### PR TITLE
Initial code around the output guardrails idea

### PR DESCRIPTION
**Pull request by cescoffier**
@geoand Because I love ruining your weekend :-)

It's non-functional right now, just some ideas of how I envisage things to work.

### CODE REVIEW

**Comment by geoand** [comment]
Merge it!!!!

File: core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/AiServicesProcessor.java [hunk 1]
```diff
@@ -1,57 +1,5 @@
 package io.quarkiverse.langchain4j.deployment;
 
-import static dev.langchain4j.exception.IllegalConfigurationException.illegalConfiguration;
```
Comment (geoand) [inline core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/AiServicesProcessor.java:3]: Not sure what happened with the import related changes...

File: core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/AiServicesProcessor.java [hunk 2]
```diff
@@ -150,9 +148,9 @@ public class AiServicesProcessor {
 
     @BuildStep
     public void nativeSupport(CombinedIndexBuildItem indexBuildItem,
-            List<AiServicesMethodBuildItem> aiServicesMethodBuildItems,
-            BuildProducer<ReflectiveClassBuildItem> reflectiveClassProducer,
-            BuildProducer<ServiceProviderBuildItem> serviceProviderProducer) {
+                              List<AiServicesMethodBuildItem> aiServicesMethodBuildItems,
```
Comment (geoand) [inline core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/AiServicesProcessor.java:151]: Seems like there is a formatting change. Probably best to check your IDE, as I don't see a good reason for this to change

File: core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/OutputGuardrail.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
+package io.quarkiverse.langchain4j.guardrails;
+
+import dev.langchain4j.data.message.AiMessage;
+import dev.langchain4j.memory.ChatMemory;
+import dev.langchain4j.rag.AugmentationResult;
+
+public interface OutputGuardrail {
```
Comment (geoand) [inline core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/OutputGuardrail.java:7]: We'll definitely need some Javadoc for this 😉

File: core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/OutputGuardrail.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+            return reprompt;
+        }
+    }
+
+
+    default void validate(AiMessage responseFromLLM) throws ValidationException {
+        validate(responseFromLLM, null, null);
+    }
+
+    void validate(AiMessage responseFromLLM, ChatMemory memory, AugmentationResult augmentationResult) throws ValidationException;
```
Comment (geoand) [inline core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/OutputGuardrail.java:34]: I would much prefer for API evolution reasons to have all these be part of a record or class instead of multiple params

File: core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/OutputGuardrails.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+@Retention(RUNTIME)
+@Target({ElementType.TYPE, ElementType.METHOD})
+public @interface OutputGuardrails {
+
+
+    /**
+     * The ordered list of guardrails to apply to the output of the model.
+     * The passed classes must be implementation of {@link OutputGuardrail}.
+     */
+    Class<?>[] value();
```
Comment (geoand) [inline core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/OutputGuardrails.java:23]:
```suggestion
    Class<? extends OutputGuardrail>[] value();
```

File: core/runtime/src/main/java/io/quarkiverse/langchain4j/runtime/aiservice/AiServiceMethodImplementationSupport.java [hunk 6]
```diff
@@ -273,9 +274,32 @@ private List<ChatMessage> messagesToSend(ChatMessage augmentedUserMessage,
             tokenUsageAccumulator = tokenUsageAccumulator.add(response.tokenUsage());
         }
 
+
+        // Guardrails chain here.
+        // TODO Should we limit the number of retry?
```
Comment (geoand) [inline core/runtime/src/main/java/io/quarkiverse/langchain4j/runtime/aiservice/AiServiceMethodImplementationSupport.java:279]: Definitely

File: core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/AiServicesProcessor.java (hunk 1 above)
Comment (cescoffier) [reply core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/AiServicesProcessor.java:3]: Don't worry, I did not run the formatter

File: core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/OutputGuardrail.java (hunk 3 above)
Comment (sberyozkin) [inline core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/OutputGuardrail.java:7]: Would it make sense to have `OutputGuardrail` implement some core interface like `OutputValidator`, and similarly for the input ? That way different types of input validators can be registered (thinking about for ex simple custom SQL query validators etc)

**Comment by cescoffier** [comment]
Start to be in a much better state with quite some tests! 

(@geoand look at how I did the tests)

File: core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/OutputGuardrail.java (hunk 3 above)
Comment (cescoffier) [reply core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/OutputGuardrail.java:19]: Done.

File: core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/OutputGuardrail.java (hunk 4 above)
Comment (cescoffier) [reply core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/OutputGuardrail.java]: Will do that tomorrow while writing the documentation.

File: core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/OutputGuardrail.java (hunk 3 above)
Comment (cescoffier) [reply core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/OutputGuardrail.java:19]: It will not be possible due to the signature difference (including exception as input guard rail won't retry or reprompt) and potential signature conflict. Because guardrails are beans, move the logic in a bean and use it from both beans.

File: core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/OutputGuardrail.java (hunk 4 above)
Comment (cescoffier) [reply core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/OutputGuardrail.java]: Actually, just did it.

**Comment by geoand** [comment]
Cool! I'll have a look tomorrow

**Comment by cescoffier** [comment]
I hope to be able to provide examples and documentation tomorrow.

File: core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/AiServicesProcessor.java [hunk 7]
```diff
@@ -614,6 +623,33 @@ public void handleDeclarativeServices(AiServicesRecorder recorder,
...
+        for (AiServicesMethodBuildItem method : methods) {
+            List<String> list = method.getGuardrails();
+            for (String cn : list) {
+                if (synthesisFinished.beanStream().withBeanType(DotName.createSimple(cn)).isEmpty()) {
+                    errors.produce(new ValidationPhaseBuildItem.ValidationErrorBuildItem(
+                            new DeploymentException("Missing guardrail bean: " + cn)));
+                }
+            }
+        }
+    }
```
Comment (geoand) [inline core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/AiServicesProcessor.java:651]: Doesn't Arc automatically do this?

Comment (cescoffier) [reply core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/AiServicesProcessor.java:651]: You get a runtime error when you try to access it (we resolve it using a programmatic lookup). Thus, it does not give you the error at build time.

Comment (geoand) [reply core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/AiServicesProcessor.java:651]: Interesting

Comment (cescoffier) [reply core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/AiServicesProcessor.java:651]: (remove that code, tests will fail :-))

**Comment by cescoffier** [comment]
@geoand ready for official review

File: docs/modules/ROOT/pages/guardrails.adoc [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+
+== Implementing Output Guardrails
+
+Output guardrails are _functions_ invoked once the LLM has produced its output.
+They are implemented as CDI beans and must implement the `io.quarkiverse.langchain4j.guardrails.OutputGuardrail` interface.
+
+Output guardrails can receive:
+
+- The output of the LLM
+- The output of the LLM, the current memory, and the augmentation results (the data chunks that were added to the prompt in the context of a RAG)
```
Comment (geoand) [inline docs/modules/ROOT/pages/guardrails.adoc:22]: Is the first line meant to be something else?

**Review by geoand — APPROVED** [review]

File: docs/modules/ROOT/pages/guardrails.adoc (hunk 8 above)
Comment (cescoffier) [reply docs/modules/ROOT/pages/guardrails.adoc:22]: there are 2 validate methods, one simply taking only the response of the LLM and another one taking a record for more context.

Let me rephrase (actually there is a section about that below)
